core/src/commonMain/kotlin/SharedRepeatableAction.kt

```diff
@@ -0,0 +1,79 @@
+package com.juul.kable
+
+import kotlinx.atomicfu.locks.reentrantLock
+import kotlinx.atomicfu.locks.withLock
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.async
+import kotlinx.coroutines.cancelAndJoin
+import kotlinx.coroutines.job
+import kotlin.coroutines.CoroutineContext
+
+/**
+ * A mechanism for launching and awaiting a shared action ([Deferred]) repeatedly.
+ *
+ * The [action] is started by calling [await]. Subsequent calls to [await] will return the same
+ * (i.e. shared) [action] until cancelled or failure occurs.
+ *
+ * The [action] is passed a [scope][CoroutineScope] that can be used to spawn coroutines that can
+ * outlive the [action]. [await] will continue to return the same action until either
+ * [cancelled][reset], or a failures occurs in either the [action] or any coroutines spawned from
+ * the [scope][CoroutineScope] provided to [action].
+ *
+ * An exception thrown from [action] will cancel any coroutines spawned from the
+ * [scope][CoroutineScope] that was provided to the [action].
+ *
+ * Calling [reset] or [resetAndJoin] will cancel the [action] and any coroutines created from the
+ * [scope][CoroutineScope] provided to the [action]. A subsequent call to [await] will then start
+ * the [action] again.
+ */
+internal class SharedRepeatableAction<T>(
+    private val coroutineContext: CoroutineContext,
+    private val action: suspend (scope: CoroutineScope) -> T,
+) {
+
+    private class Jobs<T>(
+        val root: Job,
+        val action: Deferred<T>,
+        var isCancelled: Boolean = false,
+    )
+
+    private var jobs: Jobs<T>? = null
```

Comment:
Nit: `Jobs` feels kinda collection-y, while the structure of this class is fixed at exactly 2 jobs (and even then, it's 1 parent and 1 child). `State` might be less confusing, even though it's a little generic.